File: controllers/hooks/functions/x_fetch/core/currency.py
from controllers.core_functions.core import Core
from controllers.utils import Utils
from controllers.utils.dates import Dates
utils = Utils()
pp = utils.pretty_print


def get_forex_currencies(dbms, object):
    pp("hehehehehehhehehehehehhehehe")
    currencies = []
    reporting = "ZMW"
    # settings = dbms.system_settings
    # reporting_currency = settings.linked_fields.default_currency
    # reporting = {
    #     "name": reporting_currency.name,
    #     "symbol": reporting_currency.symbol,
    #     "country": reporting_currency.country,
    #     "country_code": settings.linked_fields.default_country.code
    # }
    # forex_currencies = dbms.get_list("Currency", filters={"include_in_forex_table":1 }, fetch_linked_fields=True, page_size=12,user=object.user)
    # if forex_currencies.status != utils.ok:
    #     forex_currencies = dbms.get_list("Currency", page_size=12,user=object.user)
    #     if forex_currencies.status != utils.ok:
    #         return forex_currencies
    # for fx_c in forex_currencies.data.rows:
    #     country_code = ''
    #     country = dbms.get_doc("Country",fx_c.country,user=object.user)
    #     if country.status == utils.ok:
    #         country_code = country.data.code
    #     currencies.append({
    #         "name":fx_c.name,
    #         "symbol":fx_c.symbol,
    #         "country":fx_c.country,
    #         "country_code": country_code
    #     })
    return utils.respond(utils.ok, {"reporting_currency":reporting, "trade_currencies": currencies})
        
        
def get_trade_rates(dbms,object):
    pp("hehehehehehhehehehehehhehehe")
    rates = object.body
    xchange_rates = []
    if rates and len(rates.data) > 0:
        for rate in rates.data:
            name = rate.name
            from_currency = rate.from_currency
            to_currency = rate.to_currency
            x_rate = dbms.get_doc("Exchange_Rate", name,user=object.user)
            data = None
            if x_rate.status == utils.ok:
                data = x_rate.data
                xchange_rates.append({
                    "name": data.name,
                    "from_currency": data.from_currency,
                    "to_currency": data.to_currency,
                    "rate": data.rate,
                    "inverse": data.inverse
                })
            else:
                x_rate = dbms.get_list("Exchange_Rate",filters={"from_currency": to_currency,"to_currency": from_currency},page_size=1,user=object.user)
                if x_rate.status == utils.ok:
                    d = x_rate.data.rows[0]
                    new_exchange_rate = {
                        "name": rate.name,
                        "from_currency":rate.from_currency,
                        "to_currency":rate.to_currency,
                        "rate": d.inverse * 1,
                        "inverse": 1 / d.inverse,
                        "status":"Active"
                    }
                    new_rate = dbms.create("Exchange_Rate", utils.from_dict_to_object(new_exchange_rate),object.user)
                    xchange_rates.append(new_exchange_rate)
                else:
                    new_exchange_rate = {
                        "name": rate.name,
                        "from_currency": rate.from_currency,
                        "to_currency": rate.to_currency,
                        "rate": 1,
                        "inverse": 1,
                        "status": "Active"
                    }
                    new_rate = dbms.create("Exchange_Rate", utils.from_dict_to_object(new_exchange_rate),object.user)
                    xchange_rates.append(new_exchange_rate)
        return utils.respond(utils.ok, xchange_rates)


    
def update_exchange_rate(dbms,object):
    pass
    # core_accounting = Core_Accounting(dbms, object.user, object)
    # data = object.body.data
    # return core_accounting.update_exchange_rate()

def update_bulk_exchange_rates(dbms,object):
    pass
    # core_accounting = Core_Accounting(dbms, object.user, object)
    # data = object.body.data
    # return core_accounting.update_exchange_rate()


def get_exchange_rate(dbms,object):
    pass
    # core_accounting = Core_Accounting(dbms,object.user,object)
    # data = object.body.data
    # return core_accounting.get_exchange_rate(data.from_currency, data.to_currency)
    

def convert_currency(dbms,object):
    pass
    # core_accounting = Core_Accounting(dbms,object.user,object)
    # data = object.body.data
    # return core_accounting.convert_currency(data.from_currency, data.to_currency,data.amount)

File: static_prod/js/functions/core/currency.js

export default class Currency{
    constructor(){
        this.temp_exchange_rates = {}
    }

    async get_exchange_rate(from_currency,to_currency){
        if(from_currency === to_currency){
            return lite.connect.respond(lite.status_codes.ok, {
                from_currency: from_currency,
                to_currency: to_currency,
                rate: 1.00,
                exchange_rate: 1.00,
                inverse: 1.00,
            })
        }
        else{
            const data = {
                name: `${from_currency} - ${to_currency}`,
                from_currency: from_currency,
                to_currency: to_currency
            }
            if(this.temp_exchange_rates[data.name]){
                return {status:lite.status_codes.ok,data:this.temp_exchange_rates[data.name]}
            }
            const rate = await lite.connect.x_post("get_exchange_rate",data)
            if(rate.status === lite.status_codes.ok){
                this.temp_exchange_rates[data.name] = rate?.data
            }
            return rate
        }
    }
    async convert_currency(from_currency, to_currency, amount, convertion_rate,return_amount_only=false){
        if(!from_currency || !to_currency || !amount || amount == 0){
            if(return_amount_only){
                return amount
            }
            return lite.connect.respond(lite.status_codes.ok,{
                from_currency: from_currency,
                to_currency: to_currency,
                amount: amount,
                convertion_rate: convertion_rate
            })
        }
        if(from_currency === to_currency){
            if(return_amount_only){
                return amount
            }
            return lite.connect.respond(lite.status_codes.ok, {
                from_currency: from_currency,
                to_currency: to_currency,
                rate: 1.00,
                exchange_rate: 1.00,
                inverse: 1.00,
                converted_amount: amount
            })
        }
        else{
            const data = {
                name: `${from_currency} - ${to_currency}`,
                from_currency: from_currency,
                to_currency: to_currency,
                amount: amount,
                convertion_rate: convertion_rate
            }
            if(!this.temp_exchange_rates[data.name]){
                const rate = this.get_exchange_rate(from_currency,to_currency)
                if(rate.status === lite.status_codes.ok){
                    this.temp_exchange_rates[data.name] = rate.data
                }
                else{
                    const converted = await lite.connect.x_post("convert_currency",data)
                    if(return_amount_only){
                        return converted?.data?.converted_amount || false
                    }
                    return converted
                }
            }
            let converted = {...this.temp_exchange_rates[data.name]}
            converted.exchange_rate = converted.rate
            converted.amount = amount
            converted.converted_amount = lite.utils.fixed_decimals(converted.rate * amount, lite.defaults.currency_decimals)
            if(return_amount_only){
                return converted.converted_amount
            }
            return lite.connect.respond(lite.status_codes.ok,converted)
        }
    }

    async update_exchange_rate(from_currency,to_currency,rate){
        if(from_currency && to_currency && rate){
            const new_rates = {
                name:`${from_currency} - ${to_currency}`,
                rate:rate,
                exchange_rate:rate,
                inverse:lite.utils.fixed_decimals((1/rate),8),
                from_currency: from_currency,
                to_currency: to_currency,
            }
            const update = await lite.connect.x_post("update_exchange_rate", new_rates)
            if(update.status === lite.status_codes.ok){
                if(this.temp_exchange_rates[new_rates.name]){
                    this.temp_exchange_rates[new_rates.name].rate = rate
                }
                else{
                    this.temp_exchange_rates[update.data.name] = update.data
                }
            }
            return update
        }
    }
}